mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java

```diff
@@ -134,19 +138,15 @@ public void setName(String name) {
      * @param delimiter
      * @return list of hierarchy levels
      */
-    public List<MailboxPath> getHierarchyLevels(char delimiter) {
-        if (name == null) {
+    public List<MailboxPath> getHierarchyLevels(PathDelimiter pathDelimiter) {
+        if (Strings.isNullOrEmpty(name)) {
             return ImmutableList.of(this);
         }
-        ArrayList<MailboxPath> levels = new ArrayList<>();
-        int index = name.indexOf(delimiter);
-        while (index >= 0) {
-            final String levelname = name.substring(0, index);
-            levels.add(new MailboxPath(namespace, user, levelname));
-            index = name.indexOf(delimiter, ++index);
-        }
-        levels.add(this);
-        return levels;
+        List<String> nameParts = pathDelimiter.split(name);
+        return IntStream.range(1, nameParts.size() + 1)
```

Comment:
how about implementing that as a reducer instead ? you would just generate each Path with `last.append(current)`, right ?

Reply:
Generating the level should be a PathDelimiter responsability, no?

Ok for a StringBuilder accumulator + a side effect ;-)